### PR TITLE
Add bound callback tests

### PR DESCRIPTION
**Pull request by laradicp**

### CODE REVIEW

File: test/bound_callback_tests.jl [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        CL.Optimizer,
+        "default_optimizer" => GLPK.Optimizer,
+        "params" => CL.Params(solver = ClA.TreeSearchAlgorithm(maxnumnodes = 2))
+    )
+
+    model, x, dec = CLD.GeneralizedAssignment.model(data, coluna)
+
+    JuMP.optimize!(model)
+    # Branching constraint: x[1,1]>=1.0
+    cbdata = MathProg.PricingCallbackData(model.moi_backend.inner.re_formulation.master)
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:14]: In the pricing callback, we pass the subproblem formulation 

https://github.com/atoptima/Coluna.jl/blob/ba836391193a5f39fe39b2f1199027008b79ff19/src/Algorithm/basic/pricingcallback.jl#L26

Are you sure it's a good idea to check this after the end of the `optimize!` ?  I don't think we should guarantee that Coluna will maintain the state of the formulations after optimization ended.

Comment (laradicp) [reply test/bound_callback_tests.jl:14]: True. How can I know in which node the pricing is being made?

Comment (guimarqu) [reply test/bound_callback_tests.jl:14]: You can't but you can assume that the last time you retrieved the current bounds was in the second node.

File: test/bound_callback_tests.jl [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    # Subproblem models are created once and for all
+    # One model for each machine
+    sp_models = Dict{Int, Any}()
+    for m in data.machines
+        sp = JuMP.Model(GLPK.Optimizer)
+        @variable(sp, y[j in data.jobs], Bin)
+        @variable(sp, lb_y[j in data.jobs] >= 0)
+        @variable(sp, ub_y[j in data.jobs] >= 0)
+        @variable(sp, max_card >= 0) # this sets the maximum solution cardinality for heuristic pricing
+        @constraint(sp, card, sum(y[j] for j in data.jobs) <= max_card)
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:23]:
```suggestion
```
You can remove them, you don't use heuristic pricing

File: test/bound_callback_tests.jl [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        @constraint(sp, knp, sum(data.weight[j,m]*y[j] for j in data.jobs) <= data.capacity[m])
+        @constraint(sp, lbs[j in data.jobs], y[j] + lb_y[j] >= 0)
+        @constraint(sp, ubs[j in data.jobs], y[j] - ub_y[j] <= 0)
+        sp_models[m] = (sp, y, lb_y, ub_y, max_card)
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:27]:
```suggestion
        sp_models[m] = (sp, y, lb_y, ub_y)
```

File: test/bound_callback_tests.jl [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    end
+
+    lb = 0.0
+    ub = 1.0
+    function my_pricing_callback(cbdata)
+        machine_id = BD.callback_spid(cbdata, model)
+
+        sp, y, lb_y, ub_y, max_card = sp_models[machine_id]
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:35]:
```suggestion
        sp, y, lb_y, ub_y = sp_models[machine_id]
```

File: test/bound_callback_tests.jl [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        red_costs = [BD.callback_reduced_cost(cbdata, x[machine_id, j]) for j in data.jobs]
+
+        # Update the model
+        ## Bounds on subproblem variables
+        for j in data.jobs
+            JuMP.fix(lb_y[j], BD.callback_lb(cbdata, x[machine_id, j]), force = true)
+            JuMP.fix(ub_y[j], BD.callback_ub(cbdata, x[machine_id, j]), force = true)
+        end
+        JuMP.fix(max_card, (cbdata.stage == 1) ? length(data.jobs) : 3, force = true)
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:45]:
```suggestion
```

File: test/bound_callback_tests.jl [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        if(machine_id == 1)
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:47]:
```suggestion
        if machine_id == 1
```

File: test/bound_callback_tests.jl [hunk 7]
```diff
@@ -0,0 +1,85 @@
+function bound_callback_tests()
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:1]: Can you add a little comment to describe what you do ?

File: test/bound_callback_tests.jl (hunk 6 above)
Comment (laradicp) [reply test/bound_callback_tests.jl:47]: Hehe C if again sorry

File: test/bound_callback_tests.jl [hunk 8]
```diff
@@ -0,0 +1,84 @@
+# Test retrieval of variable bounds from pricing solver.
```
Comment (guimarqu) [inline test/bound_callback_tests.jl:1]:
```suggestion
# Test retrieval of variable bounds from pricing solver.
# We optimize the formulation with a branch-and-price and set the node limit to 2.
# We know the branching constraint (deterministic behavior) applied at the second node (x[1, 1] >= 1)
# We retrieve the current bounds of x[1, 1] in the pricing callback and we check that the last lower bound retrieved (so in the second node) is 1.
```

**Review by guimarqu — APPROVED** [review]

**Pull request merged**